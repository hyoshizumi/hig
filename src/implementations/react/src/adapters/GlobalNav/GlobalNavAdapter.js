import * as HIG from 'hig-vanilla';
import * as PropTypes from 'prop-types';

import HIGElement from '../../elements/HIGElement';
import HIGChildValidator from '../../elements/HIGChildValidator';
import createComponent from '../createComponent';

import SideNavAdapterComponent, { SideNavAdapter } from './SideNav/SideNavAdapter';
import TopNavAdapterComponent, {
  TopNavAdapter
} from './TopNav/TopNavAdapter';
import SubNavAdapterComponent, {
  SubNavAdapter
} from './SubNav/SubNavAdapter';
import Slot from '../../elements/components/GlobalNav/Slot';
import SideNav from '../../elements/components/GlobalNav/SideNav';

class GlobalNavAdapter extends HIGElement {
  constructor(initialProps) {
    super(HIG.GlobalNav, initialProps);
  }

  componentDidMount() {
    // Add any children
    if (this.sideNav) {
      this.hig.addSideNav(this.sideNav.hig);
      this.sideNav.mount();
    }

    if (this.topNav) {
      this.hig.addTopNav(this.topNav.hig);
      this.topNav.mount();
    }

    if (this.subNav) {
      this.hig.addSubNav(this.subNav.hig);
      this.subNav.mount();
    }

    if (this.initialProps.sideNavOpen) {
      this.hig.showSideNav();
    } else {
      this.hig.hideSideNav();
    }

    if (this.slot) {
      this.hig.addSlot(this.slot);
    }
  }

  createElement(ElementConstructor, props) {
    switch (ElementConstructor) {
      case TopNavAdapter:
        return new TopNavAdapter(this.hig.partials.TopNav, props);
      case SubNavAdapter:
        return new SubNavAdapter(this.hig.partials.SubNav, props);
      case SideNavAdapter:
        return new SideNavAdapter(this.hig.partials.SideNav, props);
      default:
        throw new Error(`Unknown type ${ElementConstructor.name}`);
    }
  }

  appendChild(instance, beforeChild = {}) {
    if (instance instanceof SideNavAdapter) {
      if (this.sideNav) {
        throw new Error('only one SideNav is allowed');
      } else {
        this.sideNav = instance;
        if (this.mounted) {
          this.hig.addSideNav(instance.hig);
          instance.mount();
        }
      }
    } else if (instance instanceof TopNavAdapter) {
      if (this.topNav) {
        throw new Error('only one TopNav is allowed');
      } else {
        this.topNav = instance;
        if (this.mounted) {
          this.hig.addTopNav(instance.hig);
          instance.mount();
        }
      }
    } else if (instance instanceof SubNavAdapter) {
      if (this.subNav) {
        throw new Error('only one SubNav is allowed');
      } else {
        this.subNav = instance;
        if (this.mounted) {
          this.hig.addSubNav(instance.hig);
          instance.mount();
        }
      }
    } else {
      throw new Error('unknown type');
    }
  }

  addSlot(element) {
    if (this.mounted) {
      this.hig.addSlot(element);
    } else {
      this.slot = element;
    }
  }

  insertBefore(instance) {
    this.appendChild(instance);
  }

  removeChild(instance) {
    if (instance instanceof SideNavAdapter) {
      this.sideNav = null;
    } else if (instance instanceof TopNavAdapter) {
      this.topNav = null;
    } else if (instance instanceof SubNavAdapter) {
      this.subNav = null;
    }
    instance.unmount();
  }

  commitUpdate(updatePayload, oldProps, newProp) {
    for (let i = 0; i < updatePayload.length; i += 2) {
      const propKey = updatePayload[i];
      const propValue = updatePayload[i + 1];

      switch (propKey) {
        case 'sideNavOpen': {
          propValue ? this.hig.showSideNav() : this.hig.hideSideNav();
          break;
        }
        case 'children': {
          // No-op
          break;
        }
        default: {
          console.warn(`${propKey} is unknown`);
        }
      }
    }
  }
}

const GlobalNavAdapterComponent = createComponent(GlobalNavAdapter);

GlobalNavAdapterComponent.propTypes = {
  sideNavOpen: PropTypes.bool,
  children: HIGChildValidator([
    SideNav,
    SideNavAdapterComponent,
    TopNavAdapterComponent,
    SubNavAdapterComponent,
    Slot,
  ])
};

GlobalNavAdapterComponent.__docgenInfo = {
  props: {
    sideNavOpen: {
      description: 'show or hide the SideNav'
    },

    children: {
      description: 'support adding SideNav, SubNav, or TopNav'
    }
  }
};

GlobalNavAdapterComponent.SideNav = SideNavAdapterComponent;
GlobalNavAdapterComponent.TopNav = TopNavAdapterComponent;
GlobalNavAdapterComponent.SubNav = SubNavAdapterComponent;
GlobalNavAdapterComponent.Slot = Slot;

export default GlobalNavAdapterComponent;
