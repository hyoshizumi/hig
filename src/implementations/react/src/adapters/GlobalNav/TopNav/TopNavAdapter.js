import * as PropTypes from 'prop-types';

import HIGElement from '../../../elements/HIGElement';
import HIGChildValidator from '../../../elements/HIGChildValidator';
import createComponent from '../../createComponent';

import HIGNodeList from '../../../elements/HIGNodeList';
import ProjectAccountSwitcherComponent, {
  ProjectAccountSwitcherAdapter
} from './ProjectAccountSwitcherAdapter';
import ProjectAccountSwitcher
  from '../../../elements/components/GlobalNav/TopNav/ProjectAccountSwitcher';
import ProfileComponent, {
  ProfileAdapter
} from './ProfileAdapter';
import ShortcutAdapterComponent, { ShortcutAdapter } from './ShortcutAdapter';
import HelpComponent, { HelpAdapter } from './HelpAdapter';
import SearchComponent, {
  SearchAdapter
} from './SearchAdapter';

export class TopNavAdapter extends HIGElement {
  constructor(HIGConstructor, initialProps) {
    super(HIGConstructor, initialProps);

    this.shortcuts = new HIGNodeList({
      ShortcutAdapter: {
        type: ShortcutAdapter,
        HIGConstructor: this.hig.partials.Shortcut,
        onAdd: (instance, beforeInstance) => {
          this.hig.addShortcut(instance, beforeInstance);
        }
      }
    });
  }

  componentDidMount() {
    // Add any children
    if (this.profile) {
      this.hig.addProfile(this.profile.hig);
      this.profile.mount();
    }

    if (this.projectAccountSwitcher) {
      this.hig.addProjectAccountSwitcher(this.projectAccountSwitcher.hig);
      this.projectAccountSwitcher.mount();
    }

    this.shortcuts.componentDidMount();

    if (this.help) {
      this.hig.addHelp(this.help.hig);
      this.help.mount();
    }

    if (this.search) {
      this.hig.addSearch(this.search.hig);
      this.search.mount();
    }
  }

  commitUpdate(updatePayload) {
    for (let i = 0; i < updatePayload.length; i += 2) {
      const propKey = updatePayload[i];
      const propValue = updatePayload[i + 1];

      switch (propKey) {
        case 'children': {
          // No-op
          break;
        }
        case 'logo': {
          this.hig.setLogo(propValue);
          break;
        }
        case 'logoLink': {
          this.hig.setLogoLink(propValue);
          break;
        }
        case 'onLogoClick': {
          const dispose = this._disposeFunctions.get('onLogoClickDispose');

          if (dispose) {
            dispose();
          }

          this._disposeFunctions.set(
            'onLogoClickDispose',
            this.hig.onLogoClick(propValue)
          );
          break;
        }
        default: {
          // No-op
        }
      }
    }
  }

  createElement(ElementConstructor, props) {
    switch (ElementConstructor) {
      case ProfileAdapter:
        return new ProfileAdapter(this.hig.partials.Profile, props);
      case ProjectAccountSwitcherAdapter:
        return new ProjectAccountSwitcherAdapter(
          this.hig.partials.ProjectAccountSwitcher,
          props
        );
      case HelpAdapter:
        return new HelpAdapter(this.hig.partials.Help, props);
      case ShortcutAdapter:
        return this.shortcuts.createElement(ElementConstructor, props);
      case SearchAdapter:
        return new SearchAdapter(this.hig.partials.Search, props);
      default:
        throw new Error(`Unknown type ${ElementConstructor.name}`);
    }
  }

  appendChild(instance) {
    this.requireSingleInstance(instance);
    this.checkValidChild(instance);

    if (instance instanceof ShortcutAdapter) {
      this.shortcuts.insertBefore(instance);
    } else {
      this[this.getPropertyNameFor(instance)] = instance;
      if (this.mounted) {
        this.hig[this.getFunctionNameFor(instance)](instance.hig);
        instance.mount();
      }
    }
  }

  insertBefore(instance) {
    this.appendChild(instance);
  }

  removeChild(instance) {
    if (instance instanceof ProfileAdapter) {
      this.profile = null;
    }

    instance.unmount();
  }

  requireSingleInstance(instance) {
    const requiredSingle = [
      'Profile',
      'ProjectAccountSwitcher',
      'HelpAdapter',
      'SearchAdapter'
    ];
    super.requireSingleInstance(instance, requiredSingle);
  }

  checkValidChild(instance) {
    const validInstances = [
      ProfileAdapter,
      ProjectAccountSwitcherAdapter,
      ShortcutAdapter,
      HelpAdapter,
      SearchAdapter
    ];
    if (!validInstances.includes(instance.constructor)) {
      throw new Error(
        `${instance.constructor.name} is not a valid child element of this parent.`
      );
    }
  }

  getPropertyNameFor(instance) {
    if (instance instanceof ProfileAdapter) {
      return 'profile';
    }
    if (instance instanceof ProjectAccountSwitcherAdapter) {
      return 'projectAccountSwitcher';
    }
    if (instance instanceof ShortcutAdapter) {
      return 'shortcut';
    }
    if (instance instanceof HelpAdapter) {
      return 'help';
    }
    if (instance instanceof SearchAdapter) {
      return 'search';
    }
    return null;
  }

  getFunctionNameFor(instance) {
    return `add${instance.constructor.name}`;
  }

  onHamburgerClick(callback) {
    this.hig.onHamburgerClick(callback);
  }
}

const TopNavAdapterComponent = createComponent(TopNavAdapter);

TopNavAdapterComponent.propTypes = {
  logo: PropTypes.string,
  logoLink: PropTypes.string,
  onLogoClick: PropTypes.func,
  onHamburgerClick: PropTypes.func,
  children: HIGChildValidator([
    ProfileComponent,
    ProjectAccountSwitcherComponent,
    ProjectAccountSwitcher,
    ShortcutAdapterComponent,
    HelpComponent,
    SearchComponent
  ])
};

TopNavAdapterComponent.__docgenInfo = {
  props: {
    logo: {
      description: 'sets the logo'
    },
    logoLink: {
      description: 'sets the link of the logo'
    },
    onHamburgerClick: {
      description: 'triggered when hamburger icon is clicked'
    },
    onLogoClick: {
      description: 'triggered when logo is clicked'
    }
  }
};

TopNavAdapterComponent.Profile = ProfileComponent;
TopNavAdapterComponent.Shortcut = ShortcutAdapterComponent;
TopNavAdapterComponent.Help = HelpComponent;
TopNavAdapterComponent.ProjectAccountSwitcher = ProjectAccountSwitcherComponent;
TopNavAdapterComponent.Search = SearchComponent;

export default TopNavAdapterComponent;
