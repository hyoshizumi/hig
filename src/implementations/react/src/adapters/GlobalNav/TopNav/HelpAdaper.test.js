/* eslint-disable react/prop-types, no-console */
/**
Copyright 2016 Autodesk,Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

 */
import { mount } from 'enzyme';
import * as HIG from 'hig-vanilla';
import React from 'react';

import GlobalNavAdapter from '../GlobalNavAdapter';
import TopNavAdapter from './TopNavAdapter';
import HelpAdapter from './HelpAdapter';

const Context = props => (
  <GlobalNavAdapter>
    <TopNavAdapter>
      <HelpAdapter title={props.title} link={props.link} />
    </TopNavAdapter>
  </GlobalNavAdapter>
);

function setupReactContext(props) {
  const reactContainer = document.createElement('div');
  const reactWrapper = mount(<Context {...props} />, {
    attachTo: reactContainer
  });
  return { reactWrapper, reactContainer };
}

function createHigContext(props) {
  const higContainer = document.createElement('div');

  const higNav = new HIG.GlobalNav();
  higNav.mount(higContainer);

  const higTopNav = new higNav.partials.TopNav({});
  higNav.addTopNav(higTopNav);

  const help = new higTopNav.partials.Help(props);

  higTopNav.addHelp(help);

  return { higContainer, higItem: help };
}

describe('Help', () => {
  it('renders help ', () => {
    const props = {};
    const { reactContainer } = setupReactContext(props);
    const { higContainer } = createHigContext(props);

    expect(reactContainer.firstChild.outerHTML).toMatchSnapshot();
    expect(reactContainer.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('setting and updating props', () => {
    const defaults = {
      title: 'test',
      link: 'testLink'
    };

    const nextProps = {
      title: 'newTest',
      link: 'newTestLink'
    };

    const { higContainer, higItem } = createHigContext(defaults);
    const { reactWrapper, reactContainer } = setupReactContext(defaults);

    reactWrapper.setProps(nextProps);

    higItem.setTitle(nextProps.title);
    higItem.setLink(nextProps.link);

    expect(reactContainer.firstElementChild.outerHTML).toMatchSnapshot();
    expect(reactContainer.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });
});
