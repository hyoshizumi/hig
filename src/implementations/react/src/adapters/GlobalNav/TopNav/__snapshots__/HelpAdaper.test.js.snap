// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Help renders help  1`] = `
"<div class=\\"hig__global-nav\\">
    <!--SIDENAV-->
    <div class=\\"hig__global-nav__container\\">
        <div class=\\"hig__global-nav__top-nav\\">
  <div class=\\"hig__global-nav__top-nav__group hig__global-nav__top-nav__group--left\\">
    <div class=\\"hig__global-nav__top-nav__hamburger\\">
      <div class=\\"hig__global-nav__top-nav__hamburger__hamburgericon\\"><div class=\\"hig__icon\\"><svg width=\\"16\\" height=\\"12\\" viewBox=\\"0 0 16 12\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"hamburger\\"><g transform=\\"translate(-4 -6)\\" fill=\\"#353636\\" fill-rule=\\"evenodd\\"><rect x=\\"4\\" y=\\"6\\" width=\\"16\\" height=\\"2\\"></rect><rect x=\\"4\\" y=\\"11\\" width=\\"16\\" height=\\"2\\"></rect><rect x=\\"4\\" y=\\"16\\" width=\\"16\\" height=\\"2\\"></rect></g></svg></div></div>
      <div class=\\"hig__global-nav__top-nav__hamburger__closeicon\\"><div class=\\"hig__icon\\"><svg width=\\"18\\" height=\\"18\\" viewBox=\\"0 0 18 18\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"close-hamburger\\"><g stroke=\\"#0A131C\\" stroke-width=\\"2\\" fill=\\"none\\" fill-rule=\\"evenodd\\" stroke-linecap=\\"square\\"><path d=\\"M18.754628,5.24537196 L5.24537196,18.754628\\" transform=\\"translate(-3 -3)\\"></path><path d=\\"M5.24537196,5.24537196 L18.754628,18.754628\\" transform=\\"translate(-3 -3)\\"></path></g></svg></div></div>
    </div>
    <div class=\\"hig__global-nav__top-nav__logo\\">
      <a>
        <img src=\\"\\">
      </a>
    </div>
  </div>
  <!--SEARCH-->
  <div class=\\"hig__global-nav__top-nav__spacer\\"></div>
  <!--PROJECT_ACCOUNT_SWITCHER-->
  <div class=\\"hig__global-nav__top-nav__group hig__global-nav__top-nav__group--right\\">
  <div class=\\"hig__global-nav__top-nav__shortcut\\">
  <a href=\\"#\\" title=\\"\\" class=\\"hig__global-nav__top-nav__shortcut__link\\">
    <div class=\\"hig__global-nav__top-nav__shortcut__link__icon hig__global-nav__top-nav__shortcut__link__icon--help\\">
      <div class=\\"hig__icon\\"><svg width=\\"22\\" height=\\"22\\" viewBox=\\"0 0 22 22\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"help\\"><g transform=\\"translate(-1 -1)\\" fill=\\"none\\" fill-rule=\\"evenodd\\"><rect stroke=\\"#353636\\" stroke-width=\\"1.5\\" stroke-linejoin=\\"round\\" x=\\"2.673\\" y=\\"2.673\\" width=\\"18.654\\" height=\\"18.654\\" rx=\\"9.327\\"></rect><path d=\\"M11.0622007,13.6077749 L12.6281276,13.6077749 L12.6281276,13.2731751 C12.6281276,11.7607842 14.9970939,11.6670963 14.9970939,9.84687362 L14.9970939,9.44535391 C14.9970939,8.2274108 13.9397586,7.22361152 12.1463039,7.22361152 C10.8079049,7.22361152 9.92456155,7.67866719 9.13490613,8.30771474 L9.88440958,9.29813002 C10.620529,8.78953839 11.1425047,8.50847459 11.99908,8.50847459 C12.8155035,8.50847459 13.2839431,8.93676228 13.2839431,9.59257781 L13.2839431,9.91379357 C13.2839431,11.0648167 11.0622007,11.3324965 11.0622007,13.2196392 L11.0622007,13.6077749 Z M11.8384722,16.9403885 C12.4407517,16.9403885 12.8958074,16.4719488 12.8958074,15.8830532 C12.8958074,15.3075417 12.4407517,14.825718 11.8384722,14.825718 C11.2629606,14.825718 10.7945209,15.3075417 10.7945209,15.8830532 C10.7945209,16.4719488 11.2629606,16.9403885 11.8384722,16.9403885 Z\\" fill=\\"#353636\\"></path></g></svg></div>
    </div>
  </a>
</div><!--SHORTCUT-->
  </div>
  <!--PROFILE-->
</div><!--TOPNAV-->
        <!--SUBNAV-->
        <div class=\\"hig__global-nav__slot\\">
            <!--SLOT-->
        </div>
    </div>
</div>"
`;

exports[`Help setting and updating props 1`] = `
"<div class=\\"hig__global-nav\\">
    <!--SIDENAV-->
    <div class=\\"hig__global-nav__container\\">
        <div class=\\"hig__global-nav__top-nav\\">
  <div class=\\"hig__global-nav__top-nav__group hig__global-nav__top-nav__group--left\\">
    <div class=\\"hig__global-nav__top-nav__hamburger\\">
      <div class=\\"hig__global-nav__top-nav__hamburger__hamburgericon\\"><div class=\\"hig__icon\\"><svg width=\\"16\\" height=\\"12\\" viewBox=\\"0 0 16 12\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"hamburger\\"><g transform=\\"translate(-4 -6)\\" fill=\\"#353636\\" fill-rule=\\"evenodd\\"><rect x=\\"4\\" y=\\"6\\" width=\\"16\\" height=\\"2\\"></rect><rect x=\\"4\\" y=\\"11\\" width=\\"16\\" height=\\"2\\"></rect><rect x=\\"4\\" y=\\"16\\" width=\\"16\\" height=\\"2\\"></rect></g></svg></div></div>
      <div class=\\"hig__global-nav__top-nav__hamburger__closeicon\\"><div class=\\"hig__icon\\"><svg width=\\"18\\" height=\\"18\\" viewBox=\\"0 0 18 18\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"close-hamburger\\"><g stroke=\\"#0A131C\\" stroke-width=\\"2\\" fill=\\"none\\" fill-rule=\\"evenodd\\" stroke-linecap=\\"square\\"><path d=\\"M18.754628,5.24537196 L5.24537196,18.754628\\" transform=\\"translate(-3 -3)\\"></path><path d=\\"M5.24537196,5.24537196 L18.754628,18.754628\\" transform=\\"translate(-3 -3)\\"></path></g></svg></div></div>
    </div>
    <div class=\\"hig__global-nav__top-nav__logo\\">
      <a>
        <img src=\\"\\">
      </a>
    </div>
  </div>
  <!--SEARCH-->
  <div class=\\"hig__global-nav__top-nav__spacer\\"></div>
  <!--PROJECT_ACCOUNT_SWITCHER-->
  <div class=\\"hig__global-nav__top-nav__group hig__global-nav__top-nav__group--right\\">
  <div class=\\"hig__global-nav__top-nav__shortcut\\" title=\\"newTest\\" href=\\"newTestLink\\">newTest</div><!--SHORTCUT-->
  </div>
  <!--PROFILE-->
</div><!--TOPNAV-->
        <!--SUBNAV-->
        <div class=\\"hig__global-nav__slot\\">
            <!--SLOT-->
        </div>
    </div>
</div>"
`;
